program/src/processor.rs

```diff
@@ -6365,57 +6304,901 @@ impl Processor {
-                optimal_util,
+
+            let node_bank = NodeBank::load_mut_checked(node_bank_ai, program_id)?;
+
+            // borrows must be zero
+            // deposits must be same as DustAccount
+            total_deposits += node_bank.deposits;
+            check!(node_bank.borrows.is_zero(), MangoErrorCode::InvalidAccountState)?;
+
+            // Transfer any remaining vault balance to admin owned vault, clean up token account lamports
+            // check vault was passed in
+            check!(vault_ai.key == &node_bank.vault, MangoErrorCode::InvalidVault)?;
+            let vault = Account::unpack(&vault_ai.try_borrow_data()?)?;
+            let admin_vault = Account::unpack(&admin_vault_ai.try_borrow_data()?)?;
+
+            check!(admin_vault.owner == mango_group.admin, MangoErrorCode::InvalidOwner)?;
+            let signers_seeds = gen_signer_seeds(&mango_group.signer_nonce, mango_group_ai.key);
+            invoke_transfer(
+                token_prog_ai,
+                &vault_ai,
+                &admin_vault_ai,
+                signer_ai,
+                &[&signers_seeds],
+                vault.amount,
+            )?;
+            invoke_close_token_account(
+                token_prog_ai,
+                &vault_ai,
+                admin_vault_ai,
+                signer_ai,
+                &[&signers_seeds],
+            )?;
+
+            // Close node bank, return lamports to admin
+            program_transfer_lamports(node_bank_ai, admin_ai, node_bank_ai.lamports())?;
+            sol_memset(&mut root_bank_ai.try_borrow_mut_data()?, 0, size_of::<NodeBank>());
+        }
+
+        // todo: maybe even this needs to be within some dust threshold
+        check!(
+            total_deposits == dust_account.deposits[market_index],
+            MangoErrorCode::InvalidAccountState
+        )?;
+
+        // todo: @clarkeni you think we should log this?
+        dust_account.deposits[market_index] = ZERO_I80F48;
+
+        // Close RootBank and transfer lamports to admin
+        program_transfer_lamports(root_bank_ai, admin_ai, root_bank_ai.lamports())?;
+        sol_memset(&mut root_bank_ai.try_borrow_mut_data()?, 0, size_of::<RootBank>());
+
+        // Set SpotMarketInfo to zero
+        mango_group.spot_markets[market_index] = SpotMarketInfo {
+            spot_market: Pubkey::default(),
+            maint_asset_weight: ZERO_I80F48,
+            init_asset_weight: ZERO_I80F48,
+            maint_liab_weight: ZERO_I80F48,
+            init_liab_weight: ZERO_I80F48,
+            liquidation_fee: ZERO_I80F48,
+        };
+
+        mango_group.tokens[market_index].mint = Pubkey::default();
+        mango_group.tokens[market_index].root_bank = Pubkey::default();
+        mango_group.tokens[market_index].spot_market_mode = MarketMode::Inactive;
+        Ok(())
+    }
+
+    #[inline(never)]
+    /// For a market in `SwappingSpotMarket` where we know all the OpenOrders have been closed,
```

Comment:
But we don't actually know that all openorders are closed, right? The market mode change doesn't check that, and we don't check it here either. And in fact there's no way to check for that, right?

Reply:
Yes, it's a check we have to do before we vote on the DAO proposal. It's on the list of Sanity Checks we will do in the Delist Design doc: https://www.notion.so/mango-markets/Delist-Design-03052e7719da4a30b9b9ed7ebbbbd57f#b3ad2aff7caa4df1a7303839a0b3e37b